model/test_results.go

```diff
@@ -0,0 +1,309 @@
+	}
+
+	if t.ID == "" {
+		t.ID = t.Info.ID()
+	}
+
+	updateResult, err := t.env.GetDB().Collection(testResultsCollection).UpdateOne(
+		ctx,
+		bson.M{"_id": t.ID},
+		bson.M{
+			"$set": bson.M{
+				bsonutil.GetDottedKeyName(testResultsArtifactKey, testResultsArtifactInfoTestsKey, testName): recorded,
+			},
+		},
+	)
+	grip.DebugWhen(err == nil, message.Fields{
+		"collection":   testResultsCollection,
+		"id":           t.ID,
+		"updateResult": updateResult,
+		"test_name":    testName,
+		"op":           fmt.Sprintf("set test name recorded to %v", recorded),
+	})
+	if err == nil && updateResult.MatchedCount == 0 {
+		err = errors.Errorf("could not find test results record with id %s in the database", t.ID)
+	}
+
+	return errors.Wrapf(err, "problem setting test recorded for %s", t.ID)
+}
+
+// TestResultsInfo describes information unique to a single task execution.
+type TestResultsInfo struct {
+	Project   string `bson:"project,omitempty"`
+	Version   string `bson:"version,omitempty"`
+	Variant   string `bson:"variant,omitempty"`
+	Distro    string `bson:"distro,omitempty"`
+	TaskName  string `bson:"task_name,omitempty"`
+	TaskID    string `bson:"task_id,omitempty"`
+	Execution int    `bson:"execution"`
+	Requester string `bson:"requester,omiempty"`
+	Schema    int    `bson:"schema,omitempty"`
+}
+
+var (
+	testResultsInfoProjectKey   = bsonutil.MustHaveTag(TestResultsInfo{}, "Project")
+	testResultsInfoVersionKey   = bsonutil.MustHaveTag(TestResultsInfo{}, "Version")
+	testResultsInfoVariantKey   = bsonutil.MustHaveTag(TestResultsInfo{}, "Variant")
+	testResultsInfoDistroKey    = bsonutil.MustHaveTag(TestResultsInfo{}, "Distro")
+	testResultsInfoTaskNameKey  = bsonutil.MustHaveTag(TestResultsInfo{}, "TaskName")
+	testResultsInfoTaskIDKey    = bsonutil.MustHaveTag(TestResultsInfo{}, "TaskID")
+	testResultsInfoExecutionKey = bsonutil.MustHaveTag(TestResultsInfo{}, "Execution")
+	testResultsInfoRequesterKey = bsonutil.MustHaveTag(TestResultsInfo{}, "Requester")
+	testResultsInfoSchemaKey    = bsonutil.MustHaveTag(TestResultsInfo{}, "Schema")
+)
+
+// ID creates a unique hash for a TestResults record.
+func (id *TestResultsInfo) ID() string {
+	var hash hash.Hash
+
+	if id.Schema == 0 {
+		hash = sha1.New()
+		_, _ = io.WriteString(hash, id.Project)
+		_, _ = io.WriteString(hash, id.Version)
+		_, _ = io.WriteString(hash, id.Variant)
+		_, _ = io.WriteString(hash, id.Distro)
+		_, _ = io.WriteString(hash, id.TaskName)
+		_, _ = io.WriteString(hash, id.TaskID)
+		_, _ = io.WriteString(hash, fmt.Sprint(id.Execution))
+		_, _ = io.WriteString(hash, id.Requester)
+	} else {
+		panic("unsupported schema")
+	}
+
+	return fmt.Sprintf("%x", hash.Sum(nil))
+}
+
+// TestResult describes a single test result to be stored as a BSON object in
+// some type of pail bucket storage.
+type TestResult struct {
+	TestName       string    `bson:"test_name"`
+	TestFile       string    `bson:"test_file"`
```

Comment:
what's the difference between test name and test file

my gut says it'd be good to also have the taskname here just for sanity later, but I'm negotiable.

Reply:
The test file is the name where the actual test data is written to on the host. Not sure if it is absolutely necessary but it is part of the _id in the current historical test stats.